src/markov.rs

```diff
@@ -30,99 +31,108 @@ impl Key for Markov {
+                let count = self.starting_words.entry(previous_word.clone()).or_insert(0);
                 *count += 1;
             }
+
             None => {
-                // Message is empty
+                println!("Empty message");
                 return;
             }
         }
 
-        // Get second word
-        let mut next: String;
+        let mut next_word: String;
         match words.next() {
-            Some(s) => {
-                next = String::from(s);
-                self.associate(prev.clone(), next.clone());
+            Some(word) => {
+                next_word = String::from(word);
+                self.associate(previous_word.clone(), next_word.clone());
             }
+
             None => {
-                self.associate(prev.clone(), String::from(""));
+                self.associate(previous_word.clone(), String::from(""));
                 return;
             }
         }
 
         for word in words {
-            // TODO strip punctuation and stuff
-            prev = next;
-            next = String::from(word);
-            self.associate(prev.clone(), next.clone());
+            previous_word = next_word.clone();
+            next_word = String::from(word);
+            self.associate(previous_word, next_word.clone());
         }
-        self.associate(next, String::from(""));
+        self.associate(next_word.clone(), String::from(""));
     }
 
     #[inline]
-    fn associate(&mut self, prev: String, next: String) {
-        let probs = self.assocs.entry(prev).or_insert_with(HashMap::new);
-        let count = probs.entry(next).or_insert(0);
+    fn associate(&mut self, previous_word: String, next_word: String) {
+        let probability = self.word_map.entry(previous_word).or_insert_with(HashMap::new);
+        let count = probability.entry(next_word).or_insert(0);
         *count += 1;
     }
 
-    pub fn generate(&self, length: u32) -> Option<String> {
+    pub fn generate(&self, length: u32, starting_word: Option<&String>) -> Option<String> {
         let mut rng = thread_rng();
 
-        // Get starting word
         let mut word: &String;
-        match roulette_wheel(&self.start, &mut rng) {
-            Some(x) => {
-                word = x;
+
+        match starting_word {
+            Some(start) => {
+                word = start;
             }
+
             None => {
-                // Markov chain is empty
-                return None;
+                match get_next_word(&self.starting_words, &mut rng) {
+                    Some(start) => {
+                        word = start;
+                    }
+
+                    None => {
+                        return None;
+                    }
+                }
             }
         }
 
         let mut result = String::new();
+
         for _ in 0..length {
-            match self.assocs.get(word) {
-                Some(probs) => {
+            match self.word_map.get(word) {
+                Some(probability) => {
```

Comment:
The name `probability` implies that this is one chance, like a percentage.
Rather this is a _map_ of probabilities, and thus should be plural.

Seriously, why does every variable have to be so long?